libraries/platforms/dromajo-vcs/software/src/argp/argp-parse.c

```diff
@@ -0,0 +1,992 @@
+          && strcmp (parser->state.argv[parser->state.next - 1], QUOTE)
+          == 0)
+        /* Not only is this the end of the options, but it's a
+         `quoted' region, which may have args that *look* like
+         options, so we definitely shouldn't try to use getopt past
+         here, whatever happens.  */
+        parser->state.quoted = parser->state.next;
+    }
+    else if (opt == KEY_ERR && optopt != KEY_END)
+    /* KEY_ERR can have the same value as a valid user short
+    option, but in the case of a real error, getopt sets OPTOPT
+    to the offending character, which can never be KEY_END.  */
+    {
+      *arg_ebadkey = 0;
+      return EBADKEY;
+    }
+  }
+  else
+    opt = KEY_END;
+
+  if (opt == KEY_END)
+  {
+    /* We're past what getopt considers the options.  */
+    if (parser->state.next >= parser->state.argc
+      || (parser->state.flags & ARGP_NO_ARGS))
+    /* Indicate that we're done.  */
+    {
+      *arg_ebadkey = 1;
+      return EBADKEY;
+    }
+    else
+    /* A non-option arg; simulate what getopt might have done.  */
+    {
+      opt = KEY_ARG;
+      optarg = parser->state.argv[parser->state.next++];
+    }
+  }
+
+  if (opt == KEY_ARG)
+    /* A non-option argument; try each parser in turn.  */
+    err = parser_parse_arg (parser, optarg);
+  else
+    err = parser_parse_opt (parser, opt, optarg);
+
+  if (err == EBADKEY)
+    *arg_ebadkey = (opt == KEY_END || opt == KEY_ARG);
+
+  return err;
+}
+
+/* Parse the options strings in ARGC & ARGV according to the argp in ARGP.
+   FLAGS is one of the ARGP_ flags above.  If END_INDEX is non-NULL, the
+   index in ARGV of the first unparsed option is returned in it.  If an
+   unknown option is present, EINVAL is returned; if some parser routine
+   returned a non-zero value, it is returned; otherwise 0 is returned.  */
+error_t
+__argp_parse (const struct argp *argp, int argc, char **argv, unsigned flags,
+        int *end_index, void *input)
+{
+  error_t err;
+  struct parser parser;
+
+  /* If true, then err == EBADKEY is a result of a non-option argument failing
+     to be parsed (which in some cases isn't actually an error).  */
+  int arg_ebadkey = 0;
+
+  // Sripathi:
+  // Forward declare the struct pointers since malloc is being used instead of alloca
+  struct argp_child *child;
+  struct argp *top_argp;
+  if (! (flags & ARGP_NO_HELP))
+  /* Add our own options.  */
+  {
+    // Sripathi:
+    // Originally alloca was used but since newlib does not implement that
+    // we are switching over to malloc which is a lot slower
+    child = (struct argp_child *) malloc (4 * sizeof (struct argp_child));
+    top_argp = (struct argp *) malloc (sizeof (struct argp));
+
+    /* TOP_ARGP has no options, it just serves to group the user & default
+    argps.  */
+    memset (top_argp, 0, sizeof (*top_argp));
+    top_argp->children = child;
+
+    memset (child, 0, 4 * sizeof (struct argp_child));
+
+    if (argp)
```

Comment:
in this block of code, `child` is being post incremented, which means you may be freeing an invalid pointer

Reply:
what you should do is save the original malloc'd pointer to some variable, say `child_save_ptr` or something, and then free that.

Reply:
Thanks! I think this solved the problem I have been having.